src/containers.c

```diff
@@ -0,0 +1,304 @@
+            THashData *n = h->next;
+            if (ContainerValueRangeTimeout(h->data, ts)) {
+                /* remove from the hash */
+                if (h->prev != NULL)
+                    h->prev->next = h->next;
+                if (h->next != NULL)
+                    h->next->prev = h->prev;
+                if (hb->head == h)
+                    hb->head = h->next;
+                if (hb->tail == h)
+                    hb->tail = h->prev;
+                h->next = NULL;
+                h->prev = NULL;
+                ContainerUrlRange *c = h->data;
+                FileCloseFile(c->files, NULL, 0, FILE_TRUNCATED);
+                //TODOask can we log it somehow ?
+                RangeContainerFree(c);
+                THashDataMoveToSpare(ContainerUrlRangeList.ht, h->data);
+            }
+            h = n;
+        }
+        HRLOCK_UNLOCK(hb);
+    }
+
+    return cnt;
+}
+
+void * ContainerUrlRangeGet(const uint8_t * key, size_t keylen, struct timeval *ts) {
+    ContainerUrlRange lookup;
+    // cast so as not to have const in the structure
+    lookup.key = (uint8_t *) key;
+    lookup.len = keylen;
+    struct THashDataGetResult res = THashGetFromHash(ContainerUrlRangeList.ht, &lookup);
+    if (res.data) {
+        //nothing more to do if (res.is_new)
+        ContainerUrlRangeUpdate(res.data->data, ts->tv_sec + ContainerUrlRangeList.timeout);
+        THashDecrUsecnt(res.data);
+        THashDataUnlock(res.data);
+        return res.data->data;
+    }
+    return NULL;
+}
+
+int ContainerUrlRangeSetRange(ContainerUrlRange *c, uint64_t start, uint64_t end, uint64_t total) {
+    if (total > c->totalsize) {
+        //TODOask add checks about totalsize remaining the same
+        c->totalsize = total;
+    }
+    if (start == c->files->tail->size) {
+        // easy case : append to current file
+        return 0;
+    } else if (start < c->files->tail->size) {
+        //skip first overlap
+        c->toskip = c->files->tail->size - start;
+        return 0;
+    }
+    // else {
+    // insert range in ordered linked list, if we have enough memcap
+    uint64_t buflen = end-start+1;
+    if (!(THASH_CHECK_MEMCAP(ContainerUrlRangeList.ht, buflen))) {
+        //TODOask release memory for others RangeContainerFree(c);
+        // skips this range
+        c->toskip = buflen;
+        return -1;
+    }
+    (void) SC_ATOMIC_ADD(ContainerUrlRangeList.ht->memuse, buflen);
+    RangeContainer * range = SCCalloc(1, sizeof(RangeContainer));
+    range->buffer = SCMalloc(buflen);
+    range->buflen = buflen;
+    range->start = start;
+
+    if (c->ranges == NULL || range->start < c->ranges->start) {
+        range->next = c->ranges;
+        c->ranges = range;
+    } else {
+        RangeContainer * next = c->ranges;
+        while (next->next != NULL && range->start >= next->next->start) {
+            next = next->next;
+        }
+        //TODOask merge contiguous ie if (range->start == next->start + next->buflen) {
```

Comment:
This code is simple
Another option is we could deal with overlaps here, that is sooner and maybe save some memory...